src/compiler/checker.ts

```diff
@@ -17068,6 +17068,14 @@ namespace ts {
                         return !tp.isThisType && isPartOfTypeNode(node) && maybeTypeParameterReference(node) &&
                             getTypeFromTypeNodeWorker(node as TypeNode) === tp; // use worker because we're looking for === equality
                     case SyntaxKind.TypeQuery:
+                        const entityName = (node as TypeQueryNode).exprName;
+                        const firstIdentifierDeclaration = getSymbolAtLocation(getFirstIdentifier(entityName))?.valueDeclaration;
+                        if (firstIdentifierDeclaration) {
+                            const foundSymbol = resolveName(firstIdentifierDeclaration, tp.symbol.escapedName, SymbolFlags.Type, undefined, undefined, false);

```

Comment:
Question: is using `resolveName` and checking if `foundSymbol !== tp.symbol` enough to guarantee the type parameter (`tp`) is not in scope where the first identifier is declared?

Reply:
Probably not. I think you can shadow the type variable with nested functions, like
```
function f<T>() {
  const a: T = null as any;
  function g<T>() {
    type Q = typeof a; // `T` here is the inner `T` on `g`, but the outer T is still referenced by this typeof
  }
}
```

Reply:
(While you can trivially alias `T` in the outer scope to keep it accessible)

Reply:
I rewrote it to use a duplicate-allowing version of `getSymbolsInScope`. Do you think this looks correct now?

Reply:
I don't think this strategy of looking up the symbol is going to work. In a `typeof x.y.z` it is entirely possible for the construct to reference a type parameter `T` but not actually have a type with a `symbol` property that matches `T`. For example, `x` might be of type `{ x: { y: { z: T } } }`.

Also, this code uses `getSymbolAtLocation` and `getSymbolsInScope` which are helper functions intended for the language service. In particular, `getSymbolAtLocation` has fuzzy semantics that just tries its best to come up with a symbol, and `getSymbolInScope` looks to be a rather expensive operation.

I think the right approach for determining whether `typeof x.y.z` possibly references `T` is to determine if the _location_ of the declaration of `x` falls outside the scope of that `T`. Specifically, assuming `T` is a type parameter declared in a function declaration, if a walk of the parent chain from the declaration of `x` doesn't contain that function declaration then I think it is safe to say `x` can't possibly reference `T`.

Reply:
@ahejlsberg not sure I followed the first part, what I was trying to implement was to see if at the declaration of `x.y.z`, the symbol for the given type parameter, say, `T`, was in scope.

In any case, I think what I was trying to do with `getSymbolsInScope` in the end is the same as your suggestion of walking up the declaration of `x.y.z` to see if it's within the scope of `T`, so I'm now trying to implement that. One question I have though: what should I use instead of `getSymbolAtLocation`? Since I need to go from the node for e.g. `x` (from `x.y.z`), to its declarations.

Reply:
If it's an identifier you've run `checkIdentifier` on, `getNodeLinks(ident).resolvedSymbol` is reliably the symbol.